src/replica/HttpAsyncClient.cc

```diff
@@ -0,0 +1,399 @@
+        case State::IN_PROGRESS: return "IN_PROGRESS";
+        case State::FINISHED: return "FINISHED";
+        case State::FAILED: return "FAILED";
+        case State::BODY_LIMIT_ERROR: return "BODY_LIMIT_ERROR";
+        case State::CANCELLED: return "CANCELLED";
+        case State::EXPIRED: return "EXPIRED";
+    }
+    throw invalid_argument(
+            "HttpAsyncClient::" + string(__func__) + " unknown state: "
+            + to_string(static_cast<int>(state)) + ".");
+}
+
+
+shared_ptr<HttpAsyncClient> HttpAsyncClient::create(
+        asio::io_service& io_service, CallbackType const& onFinish,
+        string const& method, string const& url, string const& data,
+        std::map<std::string, std::string> const& headers,
+        size_t maxResponseBodySize,
+        unsigned int expirationIvalSec) {
+    return shared_ptr<HttpAsyncClient>(new HttpAsyncClient(
+            io_service, onFinish, method, url, data, headers, maxResponseBodySize, expirationIvalSec));
+}
+
+
+HttpAsyncClient::HttpAsyncClient(
+        asio::io_service& io_service, CallbackType const& onFinish,
+        string const& method, string const& url, string const& data,
+        std::map<std::string, std::string> const& headers,
+        size_t maxResponseBodySize,
+        unsigned int expirationIvalSec)
+    :   _io_service(io_service),
+        _resolver(io_service),
+        _socket(io_service),
+        _onFinish(onFinish),
+        _method(::method2verb(method)),
+        _url(url),
+        _data(data),
+        _headers(headers),
+        _maxResponseBodySize(maxResponseBodySize),
+        _expirationIvalSec(expirationIvalSec),
+        _expirationTimer(io_service),
+        _timer(io_service) {
+}
+
+
+HttpAsyncClient::~HttpAsyncClient() {
+    _expirationTimer.cancel();
+    _timer.cancel();
+    boost::system::error_code ec;
+    _socket.shutdown(asio::ip::tcp::socket::shutdown_both, ec);
+}
+
+
+void HttpAsyncClient::start() {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::CREATED});
+    try {
+        _parsUrl(lock);
+        _prepareRequest(lock);
+        _resolve(lock);
+        if (_expirationIvalSec) {
+            _expirationTimer.expires_from_now(boost::posix_time::seconds(_expirationIvalSec));
+            _expirationTimer.async_wait(bind(&HttpAsyncClient::_expired, shared_from_this(), _1));
+        }
+        _state = State::IN_PROGRESS;
+    } catch(exception const& ex) {
+        _finish(lock, State::FAILED, ex.what());
+        throw;
+    }
+}
+
+
+void HttpAsyncClient::cancel() {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    switch (_state) {
+        case State::CREATED:
+            _finish(lock, State::CANCELLED);
+            return;
+        case State::IN_PROGRESS:
+            _finish(lock, State::CANCELLED);
+            return;
```

Comment:
The code in both `case`s is identical, consider merging.

Reply:
Good point.